xapian-letor/ranker/ranker.cc

```diff
@@ -351,19 +351,35 @@ Ranker::labelcomparer(const FeatureVector & firstfv, const FeatureVector& second
     return firstfv.get_label() > secondfv.get_label();
 }
 
-std::vector<Xapian::docid>
-Ranker::rank(const Xapian::MSet & mset, const string & model_key, const Xapian::FeatureList & flist)
+class ScoreIterator {
+    std::vector<FeatureVector>::const_iterator it;
+
+  public:
+    explicit ScoreIterator(const std::vector<FeatureVector>::const_iterator it_) : it(it_) { }
+
+    double operator*() const { return it->get_score(); }
+
+    void operator++() { ++it; }
+
+    bool operator!=(const ScoreIterator& o) { return it != o.it; }
+
+    bool operator==(const ScoreIterator& o) { return it == o.it; }
+
+    Xapian::doccount operator-(const ScoreIterator& o) { return Xapian::doccount(it - o.it); }
+};
+
+void
+Ranker::rank(Xapian::MSet & mset, const string & model_key, const Xapian::FeatureList & flist)
 {
+    if (mset.empty()) {
+	exit(1);
```

Comment:
You should not print anything from the library - you don't know anything about the program which is using the library.

If it's a GUI app, the output will likely vanish into `~/.xsession-errors` which is so full of junk from Gtk+ these days that basically nobody ever looks at it.  It's where error messages go to quietly die.

If it's a command-line program, then things aren't much better - if it's using curses or similar, just writing to stdout will corrupt the display.  If it's just using the console in a line-based way, you might actually manage to write output to the screen in a way the end user can see.  But the program should decide about that, not the library author.

Or the program may be a daemon and have closed all unwanted file descriptors on start-up.  In that case, fd 1 (which the the fd which `stdout` goes to) may have been reused, so by writing to `stdout` you may dump your output into a file the program is writing to and corrupt it.

Calling `exit(1)` is another thing we shouldn't do.  It causes the program to exit without any chance for vetoing that.  So if the program is some sort of editor and has an unsaved file loaded, you just threw away that unsaved file.

Actual error conditions in the library should throw an exception, not write output to `stdout` or `stderr`, and not call functions like `exit()` and `abort()` which cause the program to exit.  User code can catch and handle the exception in an appropriate way - perhaps showing the user a dialog box, or cleaning up properly before bailing out.  The program knows what's appropriate - we don't.

**HOWEVER**, the `MSet` being empty should not **NOT** an error condition.

It's certainly an edge case, but there's a perfectly valid reranking for an empty `MSet` - an empty reranking.  So as we keep saying, let's just return that rather than turning a case which has an obvious and reasonable behaviour into an error case.  If we make this an error case, then any user code using the xapian-letor API will need to check `mset.empty()` (or equivalent) before calling letor to rerank.  There's no need to impose that requirement - we can just give them back an empty rerank list, they can apply that to get a reranked `MSet` (which will still be empty).

If you were implementing addition, you wouldn't make adding zero to something an error, just because it's trivially doing nothing.  If you did that, then you force everyone to write:

```
if (b != 0) {
    a = a + b;
}
```

Which they'd rightly be annoyed by.